File: 卡码/KamaCoder/C++/test21.cpp
//21. 构造二叉树
#include <iostream>
using namespace std;

struct Node{
	char data;
	Node* left;
	Node* right;
	Node(int data):data(data),left(nullptr),right(nullptr){}
};

Node* createBTree(string preOrder,string inOrder,int preLeft,int preRight,int inLeft,int inRight){
	if(preLeft>preRight||inLeft>inRight) return nullptr;
	int rootIndex=inOrder.find(preOrder[preLeft]);
	Node* root=new Node(preOrder[preLeft]);
	int leftNum=rootIndex-inLeft;	//左子树节点个数
	root->left=createBTree(preOrder,inOrder,preLeft+1,preLeft+leftNum,inLeft,rootIndex-1);
	root->right=createBTree(preOrder,inOrder,preLeft+1+leftNum,preRight,rootIndex+1,inRight);
	return root;
}

void postOrder(Node* root){
	if(root->left!=nullptr) postOrder(root->left);
	if(root->right!=nullptr) postOrder(root->right);
	printf("%c",root->data);
}

int main(){
	string preOrder,inOrder;
	while(cin>>preOrder>>inOrder){
		Node* root=createBTree(preOrder,inOrder,0,preOrder.length()-1,0,inOrder.length()-1);
		postOrder(root);
		printf("\n");
	}
	return 0;
}

File: 卡码/KamaCoder/Js/test21.js
//21. 构造二叉树
const rl=require("readline").createInterface({input:process.stdin});
const iter=rl[Symbol.asyncIterator]();
const readline=async ()=>(await iter.next()).value;

const out=process.stdout;

class Node{
  constructor(data){
    this.data=data;
    this.left=null;
    this.right=null;
  }
  postOrder(){
    if(this.left!=null) this.left.postOrder();
    if(this.right!=null) this.right.postOrder();
    out.write(this.data);
  }
}

function createBTree(preOrder,inOrder,preLeft,preRight,inLeft,inRight){
  if(preLeft>preRight||inLeft>inRight) return null;
  let root=new Node(preOrder[preLeft]);
  let rootIndex=inOrder.indexOf(preOrder[preLeft]);
  let leftNum=rootIndex-inLeft;
  let rightNum=inRight-rootIndex;
  root.left=createBTree(preOrder,inOrder,preLeft+1,preLeft+leftNum,inLeft,rootIndex-1);
  root.right=createBTree(preOrder,inOrder,preLeft+leftNum+1,preRight,rootIndex+1,inRight);
  return root;
}

async function main(){
  while(line=await readline()){
    [preOrder,inOrder]=line.split(" ");
    const root=createBTree(preOrder,inOrder,0,preOrder.length-1,0,inOrder.length-1);
    root.postOrder();
    console.log();
  }
}

main();